chain/ethereum/src/ethereum_adapter.rs

```diff
@@ -1624,3 +1652,149 @@ pub(crate) fn parse_block_triggers(
+            }
+        }
+    };
+
+    // Confidence check: Did we collect all transactions for the current call triggers?
+    if transactions.len() != transaction_hashes.len() {
+        bail!("failed to find transactions in block for the given call triggers")
+    }
+
+    // We'll also need the receipts for those transactions. In this step we collect all receipts
+    // we have in store for the current block.
+    let mut receipts = chain_store
+        .transaction_receipts_in_block(&block.ptr().hash_as_h256())?
+        .into_iter()
+        .map(|receipt| (receipt.transaction_hash.clone(), receipt))
+        .collect::<BTreeMap<H256, LightTransactionReceipt>>();
+
+    // Do we have a receipt for each transaction under analysis?
+    let mut receipts_and_transactions: Vec<(&Transaction, LightTransactionReceipt)> = Vec::new();
+    let mut transactions_without_receipt: Vec<&Transaction> = Vec::new();
+    for transaction in transactions.iter() {
+        if let Some(receipt) = receipts.remove(&transaction.hash) {
+            receipts_and_transactions.push((transaction, receipt));
+        } else {
+            transactions_without_receipt.push(transaction);
+        }
+    }
+
+    // When some receipts are missing, we then try to fetch them from our client.
+    if !transactions_without_receipt.is_empty() {
+        let futures = transactions_without_receipt
+            .iter()
+            .map(|transaction| async move {
+                fetch_receipt_from_ethereum_client(&eth, &transaction.hash)
+                    .await
+                    .map(|receipt| (transaction, receipt))
+            });
+        futures03::future::try_join_all(futures.into_iter())
+            .await?
+            .into_iter()
+            .for_each(|(transaction, receipt)| {
+                receipts_and_transactions.push((transaction, receipt.into()))
+            });
+    }
+
+    // With all transactions and receipts in hand, we can evaluate the success of each transaction
+    let mut transaction_success: BTreeMap<&H256, bool> = BTreeMap::new();
+    for (transaction, receipt) in receipts_and_transactions.into_iter() {
+        transaction_success.insert(
+            &transaction.hash,
+            is_transaction_successful(&transaction, &receipt)?,
+        );
+    }
+
+    // Confidence check: Did we inspect the status of all transactions?
+    if !transaction_hashes
+        .iter()
+        .all(|tx| transaction_success.contains_key(tx))
+    {
+        bail!("Not all transactions status were inspected")
+    }
+
+    // Filter call triggers from unsuccessful transactions
+    block.trigger_data.retain(|trigger| {
+        if let EthereumTrigger::Call(call_trigger) = trigger {
+            // Unwrap: We already checked that those values exist
+            transaction_success[&call_trigger.transaction_hash.unwrap()]
+        } else {
+            // We are not filtering other types of triggers
+            true
+        }
+    });
+    Ok(block)
```

Comment:
It's important to note that a block with empty triggers might be returned from this.
I'm not sure if I should drop the whole block in those cases.